### PR TITLE
Add support for out parameters, boxed types, and returning an mutable optional ref. Fix some mismatches in the C FFI.

### PR DESCRIPTION
**Pull request by dgavedissian**
Out parameters are used quite often in the Physics code, so I decided to add support in luajit-ffi-gen rather than rewrite the Lua code.

This PR also fixes the annoying snake case warning generated when compiling `memory.rs`, and I fixed a few mismatches in the C FFI such as:
* When returning an option of a non-copyable type, like `Directory_Open`
* When returning a non-copyable type, the FFI was missing the `*`.

As both `as_ffi_string()` in `type_info.rs` and `wrap_type` in `generate.rs` are supposed to be equivalent to each other, I rewrote `as_ffi_string()` by taking the implementation of `wrap_type` and changing the Rust types to C. I find it much easier to follow now.

Generated `My_Struct.lua` from `test_impl.rs`, to show what the new features look like:
```lua
-- My_Struct -------------------------------------------------------------------
local Loader = {}

function Loader.declareType()
    ffi.cdef [[
        typedef struct My_Struct {} My_Struct;
    ]]

    return 1, 'My_Struct'
end

function Loader.defineType()
    local ffi = require('ffi')
    local libphx = require('libphx').lib
    local My_Struct

    do -- C Definitions
        ffi.cdef [[
            void          My_Struct_Func1                  (My_Struct const*);
            void          My_Struct_Func2                  (My_Struct*);
            void          My_Struct_FUNC3                  ();
            void          My_Struct_SetU32                 (My_Struct*, uint32 val);
            uint32        My_Struct_GetU32                 (My_Struct const*);
            void          My_Struct_SetF32                 (My_Struct*, float val);
            float         My_Struct_GetF32                 (My_Struct const*);
            void          My_Struct_SetStr                 (My_Struct*, cstr val);
            cstr          My_Struct_GetStr                 (My_Struct const*);
            void          My_Struct_SetData                (My_Struct*, Data const* val);
            void          My_Struct_TakeData               (My_Struct*, Data* val);
            void          My_Struct_TakeBoxedData          (My_Struct*, Data* val);
            Data*         My_Struct_GetData                (My_Struct const*);
            void          My_Struct_GetDataViaOutParam     (My_Struct const*, Data* out);
            Data const*   My_Struct_GetDataRef             (My_Struct const*);
            Data*         My_Struct_GetBoxedData           (My_Struct const*);
            Data*         My_Struct_GetDataMut             (My_Struct*);
            void          My_Struct_SetOpt                 (My_Struct*, uint32 const* val);
            uint32 const* My_Struct_GetOptU32              (My_Struct const*);
            Data const*   My_Struct_GetOptData             (My_Struct const*);
            void          My_Struct_SetOptRef              (My_Struct*, uint32 const* val);
            void          My_Struct_SetOptMut              (My_Struct*, uint32* val);
            uint8         My_Struct_RetResVal              ();
            uint8         My_Struct_RetResErr              ();
            uint8 const*  My_Struct_RetResOptVal           ();
            void          My_Struct_SetCopyable            (My_Struct*, WindowPos c);
            void          My_Struct_SetCopyableByRef       (My_Struct*, WindowPos const* c);
            void          My_Struct_SetCopyableByMutRef    (My_Struct*, WindowPos* c);
            WindowPos     My_Struct_GetCopyable            (My_Struct const*);
            void          My_Struct_GetCopyableViaOutParam (My_Struct const*, WindowPos* out);
            WindowPos     My_Struct_GetBoxedCopyable       (My_Struct const*);
        ]]
    end

    do -- Global Symbol Table
        My_Struct = {
            Func1                  = libphx.My_Struct_Func1,
            Func2                  = libphx.My_Struct_Func2,
            FUNC3                  = libphx.My_Struct_FUNC3,
            SetU32                 = libphx.My_Struct_SetU32,
            GetU32                 = libphx.My_Struct_GetU32,
            SetF32                 = libphx.My_Struct_SetF32,
            GetF32                 = libphx.My_Struct_GetF32,
            SetStr                 = libphx.My_Struct_SetStr,
            GetStr                 = libphx.My_Struct_GetStr,
            SetData                = libphx.My_Struct_SetData,
            TakeData               = libphx.My_Struct_TakeData,
            TakeBoxedData          = libphx.My_Struct_TakeBoxedData,
            GetData                = libphx.My_Struct_GetData,
            GetDataViaOutParam     = libphx.My_Struct_GetDataViaOutParam,
            GetDataRef             = libphx.My_Struct_GetDataRef,
            GetBoxedData           = libphx.My_Struct_GetBoxedData,
            GetDataMut             = libphx.My_Struct_GetDataMut,
            SetOpt                 = libphx.My_Struct_SetOpt,
            GetOptU32              = libphx.My_Struct_GetOptU32,
            GetOptData             = libphx.My_Struct_GetOptData,
            SetOptRef              = libphx.My_Struct_SetOptRef,
            SetOptMut              = libphx.My_Struct_SetOptMut,
            RetResVal              = libphx.My_Struct_RetResVal,
            RetResErr              = libphx.My_Struct_RetResErr,
            RetResOptVal           = libphx.My_Struct_RetResOptVal,
            SetCopyable            = libphx.My_Struct_SetCopyable,
            SetCopyableByRef       = libphx.My_Struct_SetCopyableByRef,
            SetCopyableByMutRef    = libphx.My_Struct_SetCopyableByMutRef,
            GetCopyable            = libphx.My_Struct_GetCopyable,
            GetCopyableViaOutParam = libphx.My_Struct_GetCopyableViaOutParam,
            GetBoxedCopyable       = libphx.My_Struct_GetBoxedCopyable,
        }

        if onDef_My_Struct then onDef_My_Struct(My_Struct, mt) end
        My_Struct = setmetatable(My_Struct, mt)
    end

    do -- Metatype for class instances
        local t  = ffi.typeof('My_Struct')
        local mt = {
            __index = {
                func1                  = libphx.My_Struct_Func1,
                func2                  = libphx.My_Struct_Func2,
                setU32                 = libphx.My_Struct_SetU32,
                getU32                 = libphx.My_Struct_GetU32,
                setF32                 = libphx.My_Struct_SetF32,
                getF32                 = libphx.My_Struct_GetF32,
                setStr                 = libphx.My_Struct_SetStr,
                getStr                 = libphx.My_Struct_GetStr,
                setData                = libphx.My_Struct_SetData,
                takeData               = libphx.My_Struct_TakeData,
                takeBoxedData          = libphx.My_Struct_TakeBoxedData,
                getData                = libphx.My_Struct_GetData,
                getDataViaOutParam     = libphx.My_Struct_GetDataViaOutParam,
                getDataRef             = libphx.My_Struct_GetDataRef,
                getBoxedData           = libphx.My_Struct_GetBoxedData,
                getDataMut             = libphx.My_Struct_GetDataMut,
                setOpt                 = libphx.My_Struct_SetOpt,
                getOptU32              = libphx.My_Struct_GetOptU32,
                getOptData             = libphx.My_Struct_GetOptData,
                setOptRef              = libphx.My_Struct_SetOptRef,
                setOptMut              = libphx.My_Struct_SetOptMut,
                setCopyable            = libphx.My_Struct_SetCopyable,
                setCopyableByRef       = libphx.My_Struct_SetCopyableByRef,
                setCopyableByMutRef    = libphx.My_Struct_SetCopyableByMutRef,
                getCopyable            = libphx.My_Struct_GetCopyable,
                getCopyableViaOutParam = libphx.My_Struct_GetCopyableViaOutParam,
                getBoxedCopyable       = libphx.My_Struct_GetBoxedCopyable,
            },
        }

        if onDef_My_Struct_t then onDef_My_Struct_t(t, mt) end
        My_Struct_t = ffi.metatype(t, mt)
    end

    return My_Struct
end

return Loader
```

### CODE REVIEW

File: engine/lib/luajit-ffi-gen/src/impl_item/generate.rs [hunk 1]
```diff
@@ -267,9 +272,16 @@ fn gen_func_body(self_ident: &Ident, method: &MethodInfo) -> TokenStream {
         };
 
         let method_call = if ty.is_option {
-            quote! {
-                #method_call
-                let Some(__res__) = __res__ else { return std::ptr::null(); };
+            if ty.is_mutable {
```
Comment (dgavedissian) [inline engine/lib/luajit-ffi-gen/src/impl_item/generate.rs:275]: @HaronK looks like you fixed the same thing in https://github.com/Limit-Theory-Redux/ltheory/pull/201/files#diff-b07fa8fd8254a410c8c9899fff6a102c0643785dac5687f7606c4508646abffeL269

File: engine/lib/luajit-ffi-gen/src/ffi_generator.rs [hunk 2]
```diff
@@ -188,153 +189,160 @@ impl FfiGenerator {
...
-        writeln!(&mut file, "{IDENT}return {}", self.module_name).unwrap();
+        writeln!(&mut module, "{IDENT}return {}", self.module_name).unwrap();
+
+        writeln!(&mut module, "end\n").unwrap();
+        writeln!(&mut module, "return Loader").unwrap();
 
-        writeln!(&mut file, "end\n").unwrap();
-        writeln!(&mut file, "return Loader").unwrap();
+        if cfg!(windows) {
+            module = module.replace("\n", "\r\n");
```
Comment (dgavedissian) [inline engine/lib/luajit-ffi-gen/src/ffi_generator.rs:344]: This is to generate the correct line endings on Windows

File: engine/lib/phx/script/ffi_gen/Directory.lua [hunk 3]
```diff
@@ -16,14 +16,14 @@ function Loader.defineType()
...
-            void       Directory_Free        (Directory*);
-            Directory* Directory_Open        (cstr path);
-            cstr       Directory_GetNext     (Directory*);
-            bool       Directory_Change      (cstr cwd);
-            bool       Directory_Create      (cstr path);
-            cstr       Directory_GetCurrent  ();
-            cstr       Directory_GetPrefPath (cstr org, cstr app);
-            bool       Directory_Remove      (cstr path);
+            void             Directory_Free        (Directory*);
+            Directory const* Directory_Open        (cstr path);
```
Comment (dgavedissian) [inline engine/lib/phx/script/ffi_gen/Directory.lua:20]: This is `*const Directory` in Rust, so this change is correct.

**Review by IllustrisJack — APPROVED** [review]
Looks great 2 me 👍🏽

**Comment by dgavedissian** [comment]
@HaronK I'm going to merge this now to make it easier for me to work on Rapier and WGPU, but if you have any comments, feel free to leave them after I've merged it (or make the changes in your branch :))

**Pull request merged**